csharp/src/Ice/OutputStream.cs

```diff
@@ -1675,557 +1576,307 @@ public void WriteClass(int tag, AnyClass? v)
-                {
-                    _current.sliceFlags |= Protocol.FLAG_IS_LAST_SLICE; // This is the last slice.
-                }
+                IceStartSlice(info.TypeId ?? "", firstSlice, null, info.CompactId);
+                firstSlice = false;
 
-                _stream.WriteByte(0); // Placeholder for the slice flags
+                // Write the bytes associated with this slice. TODO: need better write bytes API
+                var sliceBytes = new byte[info.Bytes.Count];
+                info.Bytes.CopyTo(sliceBytes, 0);
+                WriteBlob(sliceBytes);
 
-                //
-                // For instance slices, encode the flag and the type ID either as a
-                // string or index. For exception slices, always encode the type
-                // ID a string.
-                //
-                if (_current.sliceType == SliceType.ClassSlice)
-                {
-                    //
-                    // Encode the type ID (only in the first slice for the compact
-                    // encoding).
-                    //
-                    if (_encaps.format == FormatType.SlicedFormat || _current.firstSlice)
-                    {
-                        if (compactId >= 0)
-                        {
-                            _current.sliceFlags |= Protocol.FLAG_HAS_TYPE_ID_COMPACT;
-                            _stream.WriteSize(compactId);
-                        }
-                        else
-                        {
-                            int index = registerTypeId(typeId);
-                            if (index < 0)
-                            {
-                                _current.sliceFlags |= Protocol.FLAG_HAS_TYPE_ID_STRING;
-                                _stream.WriteString(typeId);
-                            }
-                            else
-                            {
-                                _current.sliceFlags |= Protocol.FLAG_HAS_TYPE_ID_INDEX;
-                                _stream.WriteSize(index);
-                            }
-                        }
-                    }
-                }
-                else
+                if (info.HasOptionalMembers)
                 {
-                    _stream.WriteString(typeId);
+                    _current.SliceFlags |= Protocol.FLAG_HAS_OPTIONAL_MEMBERS;
                 }
 
-                if ((_current.sliceFlags & Protocol.FLAG_HAS_SLICE_SIZE) != 0)
+                // Make sure to also re-write the instance indirection table.
+                // These instances will be marshaled (and assigned instance IDs) in IceEndSlice.
+                if (info.Instances.Count > 0)
                 {
-                    _stream.WriteInt(0); // Placeholder for the slice length.
+                    _current.IndirectionTable ??= new List<AnyClass>();
+                    Debug.Assert(_current.IndirectionTable.Count == 0);
+                    _current.IndirectionTable.AddRange(info.Instances);
```

Comment:
Hmm, we initialize IndirectionTable but not the IndirectionMap here... which explains why the map could be null when the table isn't. Is this actually correct to not add the instance to the map here? AFAICT, the object graph won't be preserved if we don't keep track of the value in the indirection map, sounds like a 3.7 bug.

Reply:
The IndirectionMap is a _per-slice_ helper that ensures we don't write twice the same class-instance in an indirection table. When we see a class a second time, we just write its local index.

Why would recreating the IndirectionMap help when writing unknown slices?

Reply:
Yes forget my comment... confusion between instanceMap and indirectionMap which aren't quite the same!